### PR TITLE
Simplify small InListExpr

### LINKED ISSUE #4089: Simplify small `InList` expressions
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
You can write a query like

```sql
select ... from foo where id in (4)
```

And that sql is oftentimes  made by tools that handle some number of ids;

We have a specialized InList implementation (e.g. see https://github.com/apache/arrow-datafusion/pull/4057) but for single values it is still faster to use a standard equality predicate

**Describe the solution you'd like**

As mentioned by @jackwener and @Dandandan in https://github.com/apache/arrow-datafusion/pull/4057#discussion_ we should rewrite inlist with a few elements.

We should definitely simplify `<left> IN (<expr>)` to `<left> = <expr>` as that will be better in all cases. 

**Describe alternatives you've considered**

We could potentially also rewrite `<left> IN (<expr>, <expr2>, .. <exprN>)` to `<left> = <expr> OR <left> = <expr2> OR .. <left> = <exprN>`

However, at some point the InList expression is faster to evaluate, and that break even point depends on the cost to evaluate `<left>`  . Thus I suggest we only rewrite for single value IN lists

**Additional context**
This is a good first issue because there are several examples of the code and tests to follow

You can find simplify rules here: https://github.com/apache/arrow-datafusion/blob/10e64dc013ba210ab1f6c2a3c02c66aef4a0e802/datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs#L329-L339

### PR DESCRIPTION
**Pull request by Dandandan**
# Which issue does this PR close?

Closes #4089

# Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by alamb** [comment]
Until the fix for https://github.com/apache/arrow-datafusion/issues/4100 is merged, clippy will be failing on this PR as well

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs [hunk 1]
```diff
@@ -365,6 +367,32 @@ impl<'a, S: SimplifyInfo> ExprRewriter for Simplifier<'a, S> {
                     None => lit_bool_null(),
                 }
             }
+            Expr::InList {
+                expr: _expr,
+                list,
+                negated,
+            } if list.is_empty() => lit(negated),
```
Comment (alamb) [inline datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:374]: I don't know how important this is, but this only holds if `expr` is non null 🤔 

```
postgres=# select null in (1);
 ?column? 
----------
 
(1 row)
```

However, note that you can't even make an empty inlist in postgres so I don't know how much this edge case matters

```sql
postgres=# select null in ();
ERROR:  syntax error at or near ")"
LINE 1: select null in ();
```

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs [hunk 2]
```diff
@@ -365,6 +367,32 @@ impl<'a, S: SimplifyInfo> ExprRewriter for Simplifier<'a, S> {
                     None => lit_bool_null(),
                 }
             }
+            Expr::InList {
```
Comment (alamb) [inline datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:370]:
```suggestion
            // expr IN () --> false
            // expr NOT IN () --> true
            Expr::InList {
```

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs [hunk 3]
```diff
@@ -365,6 +367,32 @@ impl<'a, S: SimplifyInfo> ExprRewriter for Simplifier<'a, S> {
...
+                expr: _expr,
+                list,
+                negated,
+            } if list.is_empty() => lit(negated),
+
+            Expr::InList {
```
Comment (alamb) [inline datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:376]:
```suggestion
            // if expr is a single column reference:
            // expr IN (A, B, ...) --> (expr = A) OR (expr = B) OR (expr = C)
            Expr::InList {
```

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs [hunk 4]
```diff
@@ -365,6 +367,32 @@ impl<'a, S: SimplifyInfo> ExprRewriter for Simplifier<'a, S> {
...
+                expr,
+                list,
+                negated,
+            } if list.len() <= THRESHOLD_INLINE_INLIST && expr.try_into_col().is_ok() => {
```
Comment (alamb) [inline datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:380]: I recommend also adding an explicit case for `list.len() == 1` (without `expr.try_into_col()`)

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs [hunk 5]
```diff
@@ -365,6 +367,32 @@ impl<'a, S: SimplifyInfo> ExprRewriter for Simplifier<'a, S> {
...
+                        .fold((*expr.clone()).not_eq(first_val), |acc, y| {
+                            (*expr.clone()).not_eq(y.clone()).and(acc)
+                        })
+                } else {
+                    list.into_iter()
+                        .skip(1)
+                        .fold((*expr.clone()).eq(first_val), |acc, y| {
+                            (*expr.clone()).eq(y.clone()).or(acc)
+                        })
+                }
```
Comment (alamb) [inline datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:394]: I think you can simplify this code using `reduce`:

```suggestion
                if negated {
                    list.into_iter()
                        .reduce(|acc, y| {
                            (*expr.clone()).not_eq(y.clone()).and(acc)
                        }).unwrap()
                } else {
                    list.into_iter()
                        .reduce(|acc, y| {
                            (*expr.clone()).eq(y.clone()).or(acc)
                        }).unwrap()
                }
```

Comment (Dandandan) [reply datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:398]: Unfortunately I don't think it's possible (this was my first try as well), as we also need to transform the first `acc` value.
Here the first acc won't be of `expr.eq(acc)` but just `acc`.

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs [hunk 6]
```diff
@@ -365,6 +367,38 @@ impl<'a, S: SimplifyInfo> ExprRewriter for Simplifier<'a, S> {
...
+                lit(negated)
+            }
+
+            // if expr is a single column reference:
+            // expr IN (A, B, ...) --> (expr = A) OR (expr = B) OR (expr = C)
+            Expr::InList {
+                expr,
+                list,
+                negated,
+            } if list.len() == 1
```
Comment (isidentical) [inline datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:386]: Took me a while to understand why the length of 1 is special-cased. Maybe we could mention that the column reference check is strictly for ensuring that we are not doing an unnecessary evaluation of the left side over and over again (so a length of 1 is always fine or if it is something simple as a column access then it is also fine).

**Review by isidentical — APPROVED** [review]
Looks great! One question I had was whether it makes sense to change `THRESHOLD_INLINE_INLIST` to be a configurable property (maybe inside `SimplifyInfo`) since this is more of a specific simplification for datafusion's physical in operator (when `n` is greater than 1) where other implementations might not have to perform as bad as they do for this smaller lists. Though it might be easier to wait until somebody actually needs to change the limit, so I think as is this PR looks perfect 💯

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs (hunk 6 above)
Comment (alamb) [reply datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:386]: I agree documenting the rationale would be super helpful

**Review by alamb — APPROVED** [review]
LGTM -- nice work @Dandandan

**Comment by alamb** [comment]
> Looks great! One question I had was whether it makes sense to change THRESHOLD_INLINE_INLIST to be a configurable proper

I would recommend putting this on `ConfigOptions` if you make it configurable

**Comment by alamb** [comment]
👏

**Pull request merged**